ndk/src/native_window.rs

```diff
@@ -126,4 +126,62 @@ impl NativeWindow {
     pub unsafe fn to_surface(&self, env: *mut JNIEnv) -> jobject {
         ffi::ANativeWindow_toSurface(env, self.ptr().as_ptr())
     }
+
+    /// Lock the window's next drawing surface for writing.
+    ///
+    /// Optionally pass the region you intend to draw into `dirty_bounds`.  When this function
+    /// returns it is updated (commonly enlarged) with the actual area the caller needs to redraw.
+    pub fn lock(
+        &self,
+        dirty_bounds: Option<&mut ffi::ARect>,
+    ) -> Result<NativeWindowBufferLockGuard> {
+        let dirty_bounds = match dirty_bounds {
+            Some(dirty_bounds) => dirty_bounds,
+            None => std::ptr::null_mut(),
+        };
+        let mut out = unsafe { std::mem::zeroed() };
+        let ret = unsafe { ffi::ANativeWindow_lock(self.ptr.as_ptr(), &mut out, dirty_bounds) };
+        status_to_io_result(ret, NativeWindowBufferLockGuard(out, self))
+    }
+}
+
+/// Lock holding the next drawing surface for writing.  It is unlocked and posted on [`drop()`].
+#[derive(Debug)]
+pub struct NativeWindowBufferLockGuard<'a>(ffi::ANativeWindow_Buffer, &'a NativeWindow);
+
+impl<'a> NativeWindowBufferLockGuard<'a> {
+    /// The number of pixels that are shown horizontally.
+    pub fn width(&self) -> usize {
+        usize::try_from(self.0.width).unwrap()
+    }
+
+    // The number of pixels that are shown vertically.
+    pub fn height(&self) -> usize {
+        usize::try_from(self.0.height).unwrap()
+    }
+
+    /// The number of _pixels_ that a line in the buffer takes in memory.
+    ///
+    /// This may be `>= width`.
+    pub fn stride(&self) -> usize {
+        usize::try_from(self.0.stride).unwrap()
+    }
+
+    /// The format of the buffer. One of [`HardwareBufferFormat`].
+    pub fn format(&self) -> HardwareBufferFormat {
+        let format = u32::try_from(self.0.format).unwrap();
+        HardwareBufferFormat::try_from(format).unwrap()
+    }
+
+    /// The actual bits.
+    pub fn bits(&mut self) -> *mut std::ffi::c_void {
```

Comment:
Maybe it would be possible to have a function that returns a mutable slice instead of a pointer?

Reply:
I had considered that, and especially in the context of `softbuffer` it is UB to cast a pointer to a safe slice, since these APIs are mostly for writing to memory.

It'll have to be `&[MaybeUninit<u8>]`.

Reply:
@notgull done this now. Perhaps it is also interesting to add API's that give access to individual scanlines, by offsetting the pointer by `stride() * num_rows *bpp` and giving the user only a slice of length `width() * bpp`? And then wrap that in an iterator again so that it becomes very natural to update pixels?

That won't help for `softbuffer` though which currently assumes contiguous memory.